networkx/algorithms/euler.py

```diff
@@ -147,3 +158,232 @@ def eulerian_circuit(G, source=None):
+
+    Notes
+    -----
+    Linear time algorithm, adapted from [1]_ and [3]_.
+    Information about euler paths in [2]_.
+    Code for Eulerian circuit in [3]_.
+
+    Important: In [1], euler path is in reverse order,
+    this implementation gives the path in correct order
+    as in [3]_ for eulerian_circuit. The distinction for
+    directed graph is in using the in_degree neighbors, not the out
+    ones. for undirected, it is using itemgetter(1) for get_vertex,
+    which uses the correct vertex for this order. Also, every graph
+    has an even number of odd vertices by the Handshaking Theorem [4]_.
+
+    References
+    ----------
+    .. [1] http://www.graph-magics.com/articles/euler.php
+    .. [2] http://en.wikipedia.org/wiki/Eulerian_path
+    .. [3] https://github.com/networkx/networkx/blob/master/networkx/algorithms/euler.py
+    .. [4] https://www.math.ku.edu/~jmartin/courses/math105-F11/Lectures/chapter5-part2.pdf
+    """
+    g = G.__class__(G)  # copy graph structure (not attributes)
+    is_directed = g.is_directed()
+
+    if is_directed:
+        degree = g.in_degree
+        edges = g.in_edges
+        get_vertex = itemgetter(0)
+    else:
+        degree = g.degree
+        edges = g.edges
+        get_vertex = itemgetter(1)
+
+    # Begin algorithm:
+    start = _find_path_start(g)
+    if not start:
+        raise nx.NetworkXError("G has no Eulerian path.")
+
+    vertex_stack = deque([start])
+    last_vertex = None
+
+    while vertex_stack:
+
+        current_vertex = vertex_stack[-1]
+        # if no neighbors:
+        if degree(current_vertex) == 0:
+            # Special case, we cannot add a None vertex to the path.
+            if last_vertex is not None:
+                yield (last_vertex, current_vertex)
+            last_vertex = current_vertex
+            vertex_stack.pop()
+        # we have neighbors, so add the vertex to the stack (2), take
+        # any of its neighbors (1) remove the edge between selected
+        # neighbor and that vertex, and set that neighbor as the
+        # current vertex (4).
+        else:
+            edge = next(edges(current_vertex))
+            vertex_stack.append(get_vertex(edge))
+            g.remove_edge(*edge)
+
+
+def eulerian_path(G):
+    """Return the egdes of an Eulerian path in G.
+
+    Check if the graph ``G'' has an Eulerian path or circuit and
+    return a generator for the edges. If no path is available, raise
+    an error.
+
+    Parameters
+    ----------
+    G: NetworkX Graph, DiGraph, MultiGraph or MultiDiGraph
+        A directed or undirected Graph or MultiGraph.
+
+    Returns
+    -------
+    edges: generator
+        A generator that produces edges in the Eulerian path.
+
+    Raises
+    ------
+    NetworkXError: If the graph does not have an Eulerian path.
+
+    Examples
+    --------
+    >>> G = nx.Graph([('W', 'N'), ('N', 'E'), ('E', 'W'), ('W', 'S'),
+                      ('S', 'E')])
```

Comment:
The recommended solution is to use continuation line markers: ...

https://wiki.python.org/moin/MultiLineStringsInDocTest

On 20 December 2015 at 13:32, Humberto G. Ortiz Zuazaga <
humberto.ortiz@upr.edu> wrote:

> Yes, the slash or putting G on a single line fixes the error. I've pushed
> the second.
>
> Thanks.
>
>
> On 20 December 2015 at 13:22, Sanket Dasgupta <notifications@github.com>
> wrote:
>
>> In networkx/algorithms/euler.py
>> <https://github.com/networkx/networkx/pull/1878#discussion_r48105796>:
>>
>> > +    G: NetworkX Graph, DiGraph, MultiGraph or MultiDiGraph
>> > +        A directed or undirected Graph or MultiGraph.
>> > +
>> > +    Returns
>> > +    -------
>> > +    edges: generator
>> > +        A generator that produces edges in the Eulerian path.
>> > +
>> > +    Raises
>> > +    ------
>> > +    NetworkXError: If the graph does not have an Eulerian path.
>> > +
>> > +    Examples
>> > +    --------
>> > +    >>> G = nx.Graph([('W', 'N'), ('N', 'E'), ('E', 'W'), ('W', 'S'),
>> > +                      ('S', 'E')])
>>
>> How do I run the doctest? I can look into the issue.
>>
>> It's there <https://travis-ci.org/networkx/networkx/builds/97977605> in
>> the travis logs, but if you want to check locally
>>
>> nosetests --verbosity=2 networkx/algorithms/euler.py --with-doctest
>>
>> from the top-level directory.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/networkx/networkx/pull/1878/files#r48105796>.
>>
>
>
>
> --
> Humberto Ortiz-Zuazaga
>



-- 
Humberto Ortiz-Zuazaga